skglm/solvers/lbfgs.py

```diff
@@ -46,25 +51,46 @@ def _solve(self, X, y, datafit, penalty, w_init=None, Xw_init=None):
             datafit.initialize(X, y)
 
         def objective(w):
-            Xw = X @ w
-            datafit_value = datafit.value(y, w, Xw)
-            penalty_value = penalty.value(w)
+            if self.fit_intercept:
+                Xw = X @ w[:-1] + w[-1]
+                datafit_value = datafit.value(y, w[:-1], Xw)
+                penalty_value = penalty.value(w[:-1])
+            else:
+                Xw = X @ w
+                datafit_value = datafit.value(y, w, Xw)
+                penalty_value = penalty.value(w)
 
             return datafit_value + penalty_value
 
         def d_jac(w):
-            Xw = X @ w
-            datafit_grad = datafit.gradient(X, y, Xw)
-            penalty_grad = penalty.gradient(w)
-
-            return datafit_grad + penalty_grad
+            if self.fit_intercept:
+                Xw = X @ w[:-1] + w[-1]
+                datafit_grad = datafit.gradient(X, y, Xw)
+                penalty_grad = penalty.gradient(w[:-1])
+                intercept_grad = datafit.intercept_update_step(y, Xw)
```

Comment:
there may be an issue here, because intercept_update_step does not compute the gradient, but a scaled version of it (it's mulliplied by the stepsize).
The safest way to do it would be to call `raw_grad().sum()`, which is equivalent to `np.ones(n_features) @ raw_grad()`, e.g. the gradient with respect to a feature full of ones.

Sounds good @Badr-MOUFAD ?

Reply:
Good catch @mathurinm,

I have a little concern because it seems that in some part of the code (`Logostic` datafit) `intercept_update_step` account for the stepsize, whereas in other parts (`Quadratic`, `Huber`, `Poisson`, ...) `intercept_update_step` evaluate the gradient